pkg/virt-launcher/notify-client/client.go

```diff
@@ -247,38 +247,50 @@ func eventCallback(c cli.Connection, domain *api.Domain, libvirtEvent libvirtEve
 		domain.ObjectMeta.DeletionTimestamp = &now
 		watchEvent := watch.Event{Type: watch.Modified, Object: domain}
 		client.SendDomainEvent(watchEvent)
-		events <- watchEvent
+		updateEvents(watchEvent, domain, events)
 	default:
 		if libvirtEvent.Event != nil {
 			if libvirtEvent.Event.Event == libvirt.DOMAIN_EVENT_DEFINED && libvirt.DomainEventDefinedDetailType(libvirtEvent.Event.Detail) == libvirt.DOMAIN_EVENT_DEFINED_ADDED {
 				event := watch.Event{Type: watch.Added, Object: domain}
 				client.SendDomainEvent(event)
-				events <- event
+				updateEvents(event, domain, events)
 			} else if libvirtEvent.Event.Event == libvirt.DOMAIN_EVENT_STARTED && libvirt.DomainEventStartedDetailType(libvirtEvent.Event.Detail) == libvirt.DOMAIN_EVENT_STARTED_MIGRATED {
 				event := watch.Event{Type: watch.Added, Object: domain}
 				client.SendDomainEvent(event)
-				events <- event
+				updateEvents(event, domain, events)
 			}
 		}
-		log.Log.Infof("7) OSINFO IN EVENT CALLBACK: %v", osInfo)
 		if interfaceStatus != nil {
 			domain.Status.Interfaces = interfaceStatus
 		}
 		if osInfo != nil {
 			domain.Status.OSInfo = *osInfo
 		}
-		if interfaceStatus != nil || osInfo != nil {
-			event := watch.Event{Type: watch.Modified, Object: domain}
-			client.SendDomainEvent(event)
-			events <- event
-		}
+
 		err := client.SendDomainEvent(watch.Event{Type: watch.Modified, Object: domain})
 		if err != nil {
 			log.Log.Reason(err).Error("Could not send domain notify event.")
 		}
 	}
 }
 
+var updateEvents = updateEventsClosure()
+
+func updateEventsClosure() func(event watch.Event, domain *api.Domain, events chan watch.Event) {
+	var firstAddEvent bool = true
+	var firstDeleteEvent bool = true
+
+	return func(event watch.Event, domain *api.Domain, events chan watch.Event) {
+		if event.Type == watch.Added && firstAddEvent {
+			firstAddEvent = false
+			events <- event
+		} else if event.Type == watch.Modified && domain.ObjectMeta.DeletionTimestamp != nil && firstDeleteEvent {
+			firstDeleteEvent = false
+			events <- event
```

Comment:
We can simplify this closure and create two instances one for add and the other or delete since caller knows if it's added or deleted

```golang
updateEventsOnFirstAdd := updateEventsOnFirstClosure()
updateEventsOnFirstDelete := updateEventsOnFirstClosure()

func updateEventsOnFirstClosure() func(event watch.Event, domain *api.Domain, events chan watch.Event) {
	isFirstEvent := true
	
	return func(event watch.Event, domain *api.Domain, events chan watch.Event) {
		if isFirstEvent := true
			isFirst = false
			events <- event
		} 
      }
}
```

Reply:
almost, on delete we need to know `DeletionTimestamp != nil`
i can create 2 different functions for it, but the helper func will be different a bit

Reply:
Well since the name with the proposed code is `updateVEntsOnFirstDelete` is not like people is going to copy paste that at a place that is not a delete.

Well this is just a proposal, we can forget about it.